tracsubtickets/api.py

```diff
@@ -144,97 +145,96 @@ def ticket_changed(self, ticket, comment, author, old_values):
         if new_parents == old_parents:
             return
 
-        db = self.env.get_db_cnx()
-        cursor = db.cursor()
+        with self.env.db_transaction as db:
+            cursor = db.cursor()
+
+            # remove old parents
+            for parent in old_parents - new_parents:
+                cursor.execute("DELETE FROM subtickets WHERE parent=%s AND child=%s",
+                               (parent, ticket.id))
+                # add a comment to old parent
+                xticket = Ticket(self.env, parent)
+                xticket.save_changes(author, _('Remove a subticket #%s.') % ticket.id)
+                tn = TicketNotifyEmail(self.env)
+                tn.notify(xticket, newticket=False, modtime=xticket['changetime'])
+
 
-        # remove old parents
-        for parent in old_parents - new_parents:
-            cursor.execute("DELETE FROM subtickets WHERE parent=%s AND child=%s",
-                           (parent, ticket.id))
-            # add a comment to old parent
-            xticket = Ticket(self.env, parent)
-            xticket.save_changes(author, _('Remove a subticket #%s.') % ticket.id)
-            tn = TicketNotifyEmail(self.env)
-            tn.notify(xticket, newticket=False, modtime=xticket['changetime'])
-
-
-        # add new parents
-        for parent in new_parents - old_parents:
-            cursor.execute("INSERT INTO subtickets VALUES(%s, %s)",
-                           (parent, ticket.id))
-            # add a comment to new parent
-            xticket = Ticket(self.env, parent)
-            xticket.save_changes(author, _('Add a subticket #%s.') % ticket.id)
-            tn = TicketNotifyEmail(self.env)
-            tn.notify(xticket, newticket=False, modtime=xticket['changetime'])
-
-        db.commit()
+            # add new parents
+            for parent in new_parents - old_parents:
+                cursor.execute("INSERT INTO subtickets VALUES(%s, %s)",
+                               (parent, ticket.id))
+                # add a comment to new parent
+                xticket = Ticket(self.env, parent)
+                xticket.save_changes(author, _('Add a subticket #%s.') % ticket.id)
+                tn = TicketNotifyEmail(self.env)
+                tn.notify(xticket, newticket=False, modtime=xticket['changetime'])
+
+            db.commit()
```

Comment:
`db.commit()` should be removed. The context manager `with self.env.db_transaction as db` performs a commit on exit (one of the main reasons for using context managers).